File: code/2.add-two-numbers.ts
// Definition for singly-linked list.
class ListNode {
    val: number;
    next: ListNode | null;
    constructor(val?: number, next?: ListNode | null) {
        this.val = val === undefined ? 0 : val;
        this.next = next === undefined ? null : next;
    }
}

function addTwoNumbers(
    l1: ListNode | null,
    l2: ListNode | null,
): ListNode | null {
    const dummyHead = new ListNode();
    let node: ListNode = dummyHead;
    let carry = 0;

    while (l1 || l2 || carry) {
        const val_1 = l1 ? l1.val : 0;
        const val_2 = l2 ? l2.val : 0;
        const sum = val_1 + val_2 + carry;

        carry = Math.floor(sum / 10);
        node.next = new ListNode(sum % 10);
        node = node.next;

        l1 = l1 ? l1.next : l1;
        l2 = l2 ? l2.next : l2;
    }

    return dummyHead.next;
}

File: code/2.add-two-numbers.rs
// Definition for singly-linked list.
#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
    pub val: i32,
    pub next: Option<Box<ListNode>>,
}

impl ListNode {
    #[inline]
    fn new(val: i32) -> Self {
        ListNode { next: None, val }
    }
}

impl Solution {
    pub fn add_two_numbers(
        l1: Option<Box<ListNode>>,
        l2: Option<Box<ListNode>>,
    ) -> Option<Box<ListNode>> {
        let (mut l1, mut l2) = (l1.as_ref(), l2.as_ref());

        let mut dummyHead = Box::new(ListNode::new(0));
        let mut node = dummyHead.as_mut();
        let mut carry: i32 = 0;

        while (l1.is_some() || l2.is_some() || carry != 0) {
            let val_1: i32 = if (l1.is_some()) { l1.unwrap().val } else { 0 };
            let val_2: i32 = if (l2.is_some()) { l2.unwrap().val } else { 0 };

            let sum: i32 = val_1 + val_2 + carry;

            carry = sum / 10;
        }

        return dummyHead.next;
    }
}
